File: maximumProductOfWordLengths.py
# problem: https://leetcode.com/problems/maximum-product-of-word-lengths/
# Runtime: 4552 ms, faster than 11.50% of Python3 online submissions for Maximum Product of Word Lengths.
# Memory Usage: 14.5 MB, less than 96.88% of Python3 online submissions for Maximum Product of Word Lengths.

from typing import List


class Solution:
    def maxProduct(self, words: List[str]) -> int:
        answer = 0
        for i in range(len(words)):
            for j in range(i + 1, len(words)):
                cond = False
                for character in words[i]:
                    if character in words[j]:
                        cond = True
                if not cond:
                    answer = max(answer, len(words[i]) * len(words[j]))
        return answer

File: maximumProductOfWordLengths.cpp
#include <bits/stdc++.h>


using namespace std;

class Solution {
public:
    int maxProduct(vector<string>& words) {
        vector<int> vals;
        
        for (string &word: words) {
            int tmp = 0;
            for (char &character: word) {
                tmp |= 1 << (character - 96);
            }
            vals.push_back(tmp);
        }
        int ans = 0, n = vals.size();
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j){
                if (!(vals[i] & vals[j])) {
                    int mult = words[i].length() * words[j].length();
                    ans = max(ans, mult);
                }
            }
        }
        
        return ans;
    }
};